File: fsharp/fsharp/Problem02.fs
module Problem02

open FParsec

type Cube =
    | Red
    | Green
    | Blue
    
type Game =
    { id: int
      subsets: (int * Cube) list list }
    
    static member fromTuple (id, subsets) =
        { id = id; subsets = subsets }

    
let pcube = choice [
    skipString "red" |>> fun () -> Red
    skipString "green" |>> fun () -> Green
    skipString "blue" |>> fun () -> Blue ]
let ppair = pint32 .>> spaces .>>. pcube
let psubset = sepBy ppair (pchar ',' >>. spaces)
let psubsets = sepBy psubset (pchar ';' >>. spaces)
let pgame = skipString "Game" >>. spaces >>. pint32
        .>> pchar ':' .>> spaces .>>. psubsets |>> Game.fromTuple
        
        
let toOption = function
    | Success (result, _, _) -> Some result
    | Failure (_, _, _) -> None
    
    
let is_valid_cube (count, cube) =
    match cube with
    | Red -> count <= 12
    | Green -> count <= 13
    | Blue -> count <= 14
    

let solve_a lines =
    lines
    |> Seq.map (run pgame)
    |> Seq.choose toOption
    |> Seq.filter (fun game -> game.subsets |> List.forall (List.forall is_valid_cube))
    |> Seq.map (fun game -> game.id)
    |> Seq.sum
    
let power (cubes: (int * Cube) list list) =
    cubes
    |> List.collect id
    |> List.groupBy snd
    |> List.map (snd >> List.map fst)
    |> List.map List.max
    |> List.fold (*) 1
    
    
let solve_b lines =
    lines
    |> Seq.map (run pgame)
    |> Seq.choose toOption
    |> Seq.map (fun game -> power game.subsets)
    |> Seq.sum

File: rust/src/problem02.rs
pub fn solve_a(lines: &mut dyn Iterator<Item = String>) -> u32 {
    let mut sum = 0;

    while let Some(line) = lines.next() {
    }

    sum
}

#[test]
fn solve_a_should_return_8_with_sample_data() {
    let mut lines = crate::utils::str_to_iter(indoc::indoc!("
        Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
        Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
        Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
        Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
        Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
    "));

    let actual = solve_a(&mut lines);

    assert_eq!(8, actual);
}

pub fn solve_b(lines: &mut dyn Iterator<Item = String>) -> u32 {
    let mut sum = 0;

    while let Some(line) = lines.next() {
    }

    sum
}

#[test]
fn solve_b_should_return_2286_with_sample_data() {
    let mut lines = crate::utils::str_to_iter(indoc::indoc!("
        Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
        Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
        Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
        Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
        Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
    "));

    let actual = solve_b(&mut lines);

    assert_eq!(2286, actual);
}